demo_callbacks.py

```diff
@@ -0,0 +1,281 @@
+# limitations under the License.
+
+from __future__ import annotations
+
+from typing import NamedTuple, Union
+
+import dash
+from dash import MATCH, ctx
+from dash.dependencies import Input, Output, State
+from dash.exceptions import PreventUpdate
+import plotly.graph_objs as go
+from plotly.subplots import make_subplots
+
+from data import DataSet
+from demo_interface import generate_problem_details_table_rows
+from src.demo_enums import SolverType
+from src.utils import draw_bar_chart, draw_accuracy_bars
+
+
+@dash.callback(
+    Output({"type": "to-collapse-class", "index": MATCH}, "className"),
+    inputs=[
+        Input({"type": "collapse-trigger", "index": MATCH}, "n_clicks"),
+        State({"type": "to-collapse-class", "index": MATCH}, "className"),
+    ],
+    prevent_initial_call=True,
+)
+def toggle_left_column(collapse_trigger: int, to_collapse_class: str) -> str:
+    """Toggles a 'collapsed' class that hides and shows some aspect of the UI.
+
+    Args:
+        collapse_trigger (int): The (total) number of times a collapse button has been clicked.
+        to_collapse_class (str): Current class name of the thing to collapse, 'collapsed' if not
+            visible, empty string if visible.
+
+    Returns:
+        str: The new class name of the thing to collapse.
+    """
+
+    classes = to_collapse_class.split(" ") if to_collapse_class else []
+    if "collapsed" in classes:
+        classes.remove("collapsed")
+        return " ".join(classes)
+    return to_collapse_class + " collapsed" if to_collapse_class else "collapsed"
+
+
+@dash.callback(
+        Output("features", "max"),
+        Output("features", "value"),
+        Output("features", "marks"),
+        inputs=[
+            Input("dataset", "value"),
+        ],
+)
+def create_features_input(data_set: str):
+    """Runs on load and any time the data set is updated.
+
+    Args:
+        data_set: The data set selected.
+
+    Returns:
+        int: The maximum value for the number of features slider.
+        int: The default value for the number of features slider.
+        dict: The marks for the slider to display min/max values.
+    """
+    data = DataSet(data_set)
+    max_feat = data.n
+
+    value = round(max_feat/2)
+
+    marks={
+        1: "1",
+        max_feat: str(max_feat),
+    }
+
+    return max_feat, value, marks
+
+
+@dash.callback(
+        Output("input-graph", "figure"),
+        inputs=[
+            Input("input-graph", "hoverData"),
+            Input("dataset", "value"),
+            Input("show-redund", "value")
+        ],
+)
+def draw_input_graph(hover_data: dict, data_set: str, show_red: bool) -> go.Figure:
+    """Runs on load and any time the data set is updated.
+
+    Args:
+        hover_data: Input information about user mouse location.
+        data_set: The data set selected.
+        show_red: If we want to see redundancy.
+
+    Returns:
+        go.Figure: A Plotly figure object.
+    """
+    
+    # Load the data set
```

Comment:
This will also help with the graph flicker as it will not run this function if the trigger is hover and show redundancy is not toggled

```suggestion
    """
    
        if ctx.triggered_id == "input-graph" and not show_red:
            raise PreventUpdate
    
    # Load the data set
```